From 64f1179f7f05323949ea222cc0ce9afa6a8766a3 Mon Sep 17 00:00:00 2001
From: Steve Loughran <stevel@apache.org>
Date: Tue, 17 May 2016 12:59:59 +0100
Subject: [PATCH 2383/2848] HADOOP-13163 Reuse pre-computed filestatus in
 Distcp-CopyMapper (Rajesh Balamohan via stevel)

(cherry picked from commit dfc062a0927a5f5c6db9957cd738a978d97b9705)

Change-Id: I396eb851ac1f6203fa39043dfbb343d0a04a3b63
---
 .../org/apache/hadoop/tools/mapred/CopyMapper.java |   10 ++--------
 1 file changed, 2 insertions(+), 8 deletions(-)

diff --git a/hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyMapper.java b/hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyMapper.java
index ab57127..1c6ffe9 100644
--- a/hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyMapper.java
+++ b/hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyMapper.java
@@ -242,7 +242,7 @@ public void map(Text relPath, CopyListingFileStatus sourceFileStatus,
         return;
       }
 
-      FileAction action = checkUpdate(sourceFS, sourceCurrStatus, target);
+      FileAction action = checkUpdate(sourceFS, sourceCurrStatus, target, targetStatus);
       if (action == FileAction.SKIP) {
         LOG.info("Skipping copy of " + sourceCurrStatus.getPath()
                  + " to " + target);
@@ -329,13 +329,7 @@ private static void incrementCounter(Context context, Counter counter,
   }
 
   private FileAction checkUpdate(FileSystem sourceFS, FileStatus source,
-      Path target) throws IOException {
-    final FileStatus targetFileStatus;
-    try {
-      targetFileStatus = targetFS.getFileStatus(target);
-    } catch (FileNotFoundException e) {
-      return FileAction.OVERWRITE;
-    }
+      Path target, FileStatus targetFileStatus) throws IOException {
     if (targetFileStatus != null && !overWrite) {
       if (canSkip(sourceFS, source, targetFileStatus)) {
         return FileAction.SKIP;
-- 
1.7.9.5

